### PR TITLE
Make stats summary tests more flexible

### LINKED ISSUE #95344: Summary API test is a bit flaky
**Which jobs are flaking**:
[Kubelet-node-master](https://k8s-testgrid.appspot.com/sig-release-master-blocking#node-kubelet-master)

**Which test(s) are flaking**:
[Summary API NodeConformance](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/summary_test.go#L41)

**Testgrid link**:
https://k8s-testgrid.appspot.com/sig-release-master-blocking#node-kubelet-master

**Reason for failure**:
Based on the execution conditions sometimes the value of `Node.SystemContainers[pods].CPU.UsageNanoCores` could be 0 instead of expected 10000

**Anything else we need to know**:
- links to go.k8s.io/triage appreciated
- links to specific failures in spyglass appreciated

### PR DESCRIPTION
**Pull request by harche**
Signed-off-by: Harshal Patil <harpatil@redhat.com>

**What type of PR is this?**

/kind failing-test
/kind flake

**What this PR does / why we need it**:
Summary API test is a bit [flaky on ubuntu right now](https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-node-kubelet/1313556070618107904), and it also [fails on CRIO/Fedora CoreOS](https://github.com/kubernetes/kubernetes/issues/95343) because the output of `/stats/summary` isn't same. 
 
**Which issue(s) this PR fixes**:

Fixes #95344, #95343

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by harche** [comment]
Flaky test issue - https://github.com/kubernetes/kubernetes/issues/95344
Test grid link of flaky tests - https://k8s-testgrid.appspot.com/sig-release-master-blocking#node-kubelet-master
Sample failed test - https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-node-kubelet/1313556070618107904
Failure on CRIO/Fedora CoreOS - https://github.com/kubernetes/kubernetes/issues/95343

File: test/e2e_node/summary_test.go
Comment (harche) [inline test/e2e_node/summary_test.go:93]: This change is related to flaky test behaviour - https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-node-kubelet/1313556070618107904

File: test/e2e_node/summary_test.go
Comment (harche) [inline test/e2e_node/summary_test.go:185]: These changes required because the code enters in `else`, https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/stats/helper.go#L72 instead of `if`, https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/stats/helper.go#L56

**Comment by harche** [comment]
Apart from fixing the [flake](https://k8s-testgrid.appspot.com/sig-release-master-blocking#node-kubelet-master), this PR also fixes the assumptions on the platform that may not be true on CRIO/Fedora Cores. Here is the sample output of `/stats/summary` from CRIO/Fedora CoreOS, https://github.com/kubernetes/kubernetes/issues/95343#issuecomment-704748920

File: test/e2e_node/summary_test.go
Comment (alejandrox1) [reply test/e2e_node/summary_test.go:93]: Could you please add a comment of some sort explaining more in detail why this change is needed?
I see you added some sample output from /stats/summary in https://github.com/kubernetes/kubernetes/issues/95343#issuecomment-704748920 and it does show some pods with 0 but why would that be the case?

I've noticed this test is flaky in fedora and ubuntu. so my question is also: is there something wrong with the configuration of the test environment in fedore/ubuntu or is there some bad assumption that went into the current version of this test?
Ultimately, why should we check that the value of `UsageNanoCores` is between 0 and 2e9 instead of 10k and 2e9?

File: test/e2e_node/summary_test.go
Comment (alejandrox1) [reply test/e2e_node/summary_test.go:185]: What's the path between `cadvisorInfoToCPUandMemoryStats` in https://github.com/kubernetes/kubernetes/blob/e99df0e5a75eb6e86123b56d53e9b7ca0fd00419/pkg/kubelet/stats/helper.go#L36

and the test itself in https://github.com/kubernetes/kubernetes/blob/e99df0e5a75eb6e86123b56d53e9b7ca0fd00419/test/e2e_node/summary_test.go#L41-L43

not seeing any calls to `cadvisorInfoToCPUandMemoryStats` in the test :thinking:

**Review by alejandrox1 — COMMENTED** [review]
Left some questions in here.
I could see how this would help the test failures we see but I am not sure if this solution is what is needed - mainly because it isn't completely clear from the test why such conditions are being tested.
Could you describe your reasoning for this patch more in depth, please?
/hold

File: test/e2e_node/summary_test.go
Comment (mmerkes) [reply test/e2e_node/summary_test.go:93]: It's unclear why 0 would be a valid state. Since it seems to be failing on the `gomega.Constantly` validation, it already passed the `gomega.Eventually` validation, meaning that the value was already greater than 10k. Is this covering up an issue? Or, is the idea that this is not the level of validation required for this API?

Looks like `UsageNanoCores` is zero when [cadvisor doesn't provide them.](https://github.com/kubernetes/kubernetes/blame/d19aec8bf1c8ca294345001d2286db578d422982/pkg/kubelet/stats/helper.go#L45)

Comment (harche) [reply test/e2e_node/summary_test.go:93]: @alejandrox1 on fedora (and probably on ubuntu too) stats provider is CRI while on COS it might be cadvisor if they use docker. 

CRI stats provider tries to estimate the value of `UsageNanoCores` because CRI runtimes do not supply nano core usages. Check this [comment](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/stats/cri_stats_provider.go#L106) on that function for more details. If any of these [conditions](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/stats/cri_stats_provider.go#L667) are met then the value of the `UsageNanoCores` for that request won't be set. 

Similarly, If [this](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/stats/cri_stats_provider.go#L533) call returns `nil`, we may never have the value for `UsageNanoCores` set. 

@mmerkes So the reason I set to 0, because for CRI stats provider the value can indeed be 0 unless the values in the cache are built up and [used in calculating](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/stats/cri_stats_provider.go#L679) the effective `UsageNanoCores`

Comment (harche) [reply test/e2e_node/summary_test.go:93]: I was just wondering, if it would help to play with the timeout values of `gomega.Constantly` and `gomega.Constantly` to see if that mitigates the issue.

File: test/e2e_node/summary_test.go
Comment (harche) [reply test/e2e_node/summary_test.go:185]: `summary_test.go` calls [getNodeSummary](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/util.go#L74) which makes http request to kubelet's `/stats/summary` API endpoint. 

This endpoint is handled by, https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/server/stats/handler.go#L221 which in tern end ups calling `cadvisorInfoToCPUandMemoryStats` via the [CRI stats provider](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/stats/cri_stats_provider.go).

Anyway, while scanning through kubelet logs, I found some more errors that may or may not be related to the fixes below, 

```
Oct 08 09:22:56 localhost kubelet[3337]: E1008 09:22:56.037312    3337 cri_stats_provider.go:377] Failed to get the info of the filesystem with mountpoint "/var/lib/containers/storage/overlay/c4b641869c0da8b82b97bd2f7833f1b2a778acf415c0d3d99a289c1426a95f19/merged": failed to get device for dir "/var/lib/containers/storage/overlay/c4b641869c0da8b82b97bd2f7833f1b2a778acf415c0d3d99a289c1426a95f19/merged": could not find device with major: 0, minor: 71 in cached partitions map.
Oct 08 09:22:56 localhost kubelet[3337]: I1008 09:22:56.042350    3337 cri_stats_provider.go:444] Unable to find network stats for sandbox "01db264e3ae0b4c2e12dbaf72ef2863760e521802ca886b00e0554b66125f754"
```
so I will separate these 2 issues. Let's focus on `UsageNanoCores` in this PR and if needed I can handle the other changes in this PR separately.

File: test/e2e_node/summary_test.go
Comment (harche) [reply test/e2e_node/summary_test.go:93]: Just to add another point, I had 5 runs of this test on Fedora CoreOS by keeping the accepted values between 10000 and 2e9, and they all succeeded. I am not able to predicatibly reproduce this so far, pretty much like the behaviour we see on [test grid](https://k8s-testgrid.appspot.com/sig-release-master-blocking#node-kubelet-master).

**Comment by harche** [comment]
Keeping this PR only focused on `UsageNanoCores`,  I have purged the other changes in this PR and will handle them in separate PR if needed.

File: test/e2e_node/summary_test.go
Comment (harche) [inline test/e2e_node/summary_test.go:93]: Relevant discussion from earlier changes, https://github.com/kubernetes/kubernetes/pull/95345#discussion_r500796556

**Comment by alejandrox1** [comment]
Additional question that came to mind @harche 

> Summary API test is a bit flaky on ubuntu right now, and it also fails on CRIO/Fedora CoreOS because the output of /stats/summary isn't same.

Did you manually run this test on fedora or is there a testgrid dashboard where this test is being executed in fedora?
If you ran manually, how many times did you run it? how many failures did you see? what was your setup?

File: test/e2e_node/summary_test.go
Comment (alejandrox1) [reply test/e2e_node/summary_test.go]: @harche could you document in someway what you said in https://github.com/kubernetes/kubernetes/pull/95345#discussion_r501630942 , please?
I think it makes sense, and this patch looks reasonable.
Plus it may be good to merge and see how the behaviour of the test changes.

Though at least for this test (just by looking at it) some of the conditions seem a tad arbitrary.
The reason behind this PR is interesting in some way and i think it would be good to leave a comment in the code detailing the why of this check.

**Review by alejandrox1 — COMMENTED** [review]
I think that with https://github.com/kubernetes/kubernetes/pull/95345#discussion_r501630942 documented in the test itself this may be good.

/triage accepted
/remove-sig testing

**Comment by harche** [comment]
@alejandrox1 I use following command to run node e2e on Fedora CoreOS,

```
make test-e2e-node REMOTE=true RUNTIME=remote IMAGE_CONFIG_FILE="/home/harshal/Downloads/image-config-cgrpv1.yaml" INSTANCE_PREFIX="harpatil" CLEANUP=false CONTAINER_RUNTIME_ENDPOINT="unix:///var/run/crio/crio.sock" FOCUS="\[HEYHOTEST\]" SKIP="\[Flaky\]|\[Slow\]|\[Serial\]" TEST_ARGS='--kubelet-flags="--cgroup-driver=systemd --cgroups-per-qos=true --cgroup-root=/ --runtime-cgroups=/system.slice/crio.service --non-masquerade-cidr=0.0.0.0/0" --extra-log="{\"name\": \"crio.log\", \"journalctl\": [\"-u\", \"crio\"]}"
```

the image config file is from, https://github.com/harche/crio-fcos-config/blob/master/image-config-cgrpv1.yaml

But it needs some more patches (PRs still not merged) to successfully launch the node e2e job on Fedora CoreOS host. 

I have run those jobs several times, but the failures due to 0 value of `UsageNanoCores` very actually very rare with FCOS host. To be precise, out of hundreds of host I have seen it just twice. So that issue is not as frequent on FCOS as on Ubuntu.

**Comment by alejandrox1** [comment]
Interesting, wondering how to follow up to actually run these tests in fedore.
One last thing to merge, @harche could you please add some comments on the test itself summarizing the reason why, please?
ref: https://github.com/kubernetes/kubernetes/pull/95345#discussion_r504057582

I think this change may be a good thing to experiment but it will also be helpful to have some comments to go along with the code to iterate on this test - otherwise the values seem somewhat arbitrary.

**Comment by harche** [comment]
@alejandrox1 I added a comment with a link that provides the information from this discussion over here. 

About running the Fedora CoreOS based tests, I have WIP PR, https://github.com/kubernetes/test-infra/pull/19342 that will add some pre-submit prow jobs. I am also working on adding some periodic jobs as well. Once those are merged in test-infra, we should be able to trigger the jobs and observe the results on testgrid for Fedora CoreOS based nodes.

**Review by alejandrox1 — APPROVED** [review]
@harche thank you for all the work and conversation here!
(excuse the delay, took some time off) but let's get this one going.
/hold cancel
/lgtm
/approve

**Comment by dashpole** [comment]
Shouldn't we only get a 0 value on the very first scrape?  This sounds like a bug, rather than something we should relax tests for.

**Comment by dashpole** [comment]
cc @SergeyKanzhelev

**Pull request merged**
